crates/red_knot_test/src/parser.rs

```diff
@@ -0,0 +1,467 @@
+        let root_section_id = sections.push(Section {
+            title: path,
+            level: 0,
+            parent_id: None,
+        });
+        Self {
+            sections,
+            files: IndexVec::default(),
+            rest: source,
+            stack: vec![root_section_id],
+            current_section_files: None,
+        }
+    }
+
+    fn parse(mut self) -> anyhow::Result<MarkdownTestSuite<'s>> {
+        self.start()?;
+        Ok(self.finish())
+    }
+
+    fn finish(mut self) -> MarkdownTestSuite<'s> {
+        self.sections.shrink_to_fit();
+        self.files.shrink_to_fit();
+
+        MarkdownTestSuite {
+            sections: self.sections,
+            files: self.files,
+        }
+    }
+
+    fn start(&mut self) -> anyhow::Result<()> {
+        while !self.rest.is_empty() {
+            if let Some(caps) = self.scan(&HEADER_RE) {
+                let header_level = caps["level"].len();
+                self.pop_sections_to_level(header_level);
+
+                // We never pop the implicit root section.
+                let parent = self.stack.last().unwrap();
+
+                let section = Section {
+                    // HEADER_RE can't match without a match for group 'title'.
+                    title: caps.name("title").unwrap().into(),
+                    level: header_level,
+                    parent_id: Some(*parent),
+                };
+
+                if self.current_section_files.is_some() {
+                    return Err(anyhow::anyhow!(
+                        "Header '{}' not valid inside a test case; parent '{}' has code files.",
+                        section.title,
+                        self.sections[*parent].title,
+                    ));
+                }
+
+                let section_id = self.sections.push(section);
+                self.stack.push(section_id);
+
+                self.current_section_files = None;
+            } else if let Some(caps) = self.scan(&CODE_RE) {
+                // We never pop the implicit root section.
+                let parent = self.stack.last().unwrap();
+
+                let mut config: FxHashMap<&'s str, &'s str> = FxHashMap::default();
+
+                if let Some(config_match) = caps.name("config") {
+                    for item in config_match.as_str().split_whitespace() {
+                        let mut parts = item.split('=');
+                        let key = parts.next().unwrap();
+                        let Some(val) = parts.next() else {
+                            return Err(anyhow::anyhow!("Invalid config item `{}`.", item));
+                        };
+                        if parts.next().is_some() {
+                            return Err(anyhow::anyhow!("Invalid config item `{}`.", item));
+                        }
+                        config.insert(key, val);
+                    }
+                }
+
+                let path = config.get("path").unwrap_or(&"test.py");
```

Comment:
```suggestion
                let path = config.get("path").copied().unwrap_or("test.py");
```

this means that `path` has type `&str` rather than `&&str`, which then allows you to make these simplifications:

```diff
@@ -220,16 +222,16 @@ impl<'s> Parser<'s> {
                 });
 
                 if let Some(current_files) = &mut self.current_section_files {
-                    if current_files.contains(*path) {
+                    if current_files.contains(path) {
                         let current = &self.sections[*self.stack.last().unwrap()];
                         return Err(anyhow::anyhow!(
                             "Test `{}` has duplicate files named `{path}`.",
                             current.title
                         ));
                     }
-                    current_files.insert(*path);
+                    current_files.insert(path);
                 } else {
-                    self.current_section_files = Some(FxHashSet::from_iter([*path]));
+                    self.current_section_files = Some(FxHashSet::from_iter([path]));
                 }
```